### PR TITLE
[binding] Reduce binding manager size

### PR DESCRIPTION
**Pull request by gjc13**
#### Problem

The `BindingManger` class consumes about 1KB of RAM.

#### Change overview

The PR stores the pending notifications as a flat array of binding table indecies. The RAM usage is reduced to about 250 bytes.

#### Testing

Build `all-cluster-app` with gn args `chip_build_libshell=true`
Run two linux all-cluster-app in different docker containers or ip namespaces.
```
chip-tool pairing ethernet 1 20202021 3840 172.17.0.2 5540
chip-tool pairing ethernet 2 20202021 3840 172.17.0.3 5540
chip-tool binding bind 1 0 1 6 2 1
```

Now in the second device's shell:

```
> switch on
> switch off
```

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

File: src/app/clusters/bindings/BindingManager.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/BindingManager.cpp:39]: Is there a reason to not make this take a cost ref to the binding table entry?  The caller has one (on stack)....  And we never use the id here except to get the entry.

File: src/app/clusters/bindings/BindingManager.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/BindingManager.cpp:54]: Is this cross-check that the caller did not mess up worth doing?  I would just assume that if it's claiming to have removed a unicast binding then it did....

File: src/app/clusters/bindings/BindingManager.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/BindingManager.cpp:79]: Wait, why is this cast OK?

File: src/app/clusters/bindings/BindingManager.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/BindingManager.cpp:99]: Shouldn't we remove the entries we notified for?

File: src/app/clusters/bindings/BindingManager.h
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/BindingManager.h:73]:
```suggestion
     * Notifies the BindingManager that a unicast binding is about to be removed from the given index.
```

File: src/app/clusters/bindings/PendingNotificationMap.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/PendingNotificationMap.cpp:27]: How big will this stack use be?

File: src/app/clusters/bindings/PendingNotificationMap.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/PendingNotificationMap.cpp:38]: So this is O(N^2) in size of the binding table.  How big is that likely to be?

File: src/app/clusters/bindings/PendingNotificationMap.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/PendingNotificationMap.cpp:64]: It's very non-obvious what this is doing and why.  Why does it matter which bindings for a peer come before others in the binding table (which this code somehow cares about)?  What notion of "LRU" is this trying to implement, exactly?

This needs either much better variable naming or much better comments or most likely both.  But I also suspect that the behavior this is implementing does not make sense.

File: src/app/clusters/bindings/PendingNotificationMap.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/PendingNotificationMap.cpp:92]: What happens when mNextToOverride exceeds kMaxPendingNotifications?

File: src/app/clusters/bindings/PendingNotificationMap.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/PendingNotificationMap.cpp:99]: How big will these stack bits get?  I guess this is trying to avoid O(N^2) behavior from doing memmove() multiple times?

File: src/app/clusters/bindings/PendingNotificationMap.h
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/PendingNotificationMap.h:48]: This does not look right to me.

Concretely, say kMaxPendingNotifications == mNumEntries == 3, and mNextToOverride == 1.

mIndex starts at 1.  Then it becomes 2.  Then it becomes 3, tests true for == mNumEntries, becomes -1, and iteration ends. But we never iterated the thing at index 0.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/clusters/bindings/PendingNotificationMap.cpp
Comment (gjc13) [reply src/app/clusters/bindings/PendingNotificationMap.cpp:38]: Typically 10 to 20 so still acceptable.

File: src/app/clusters/bindings/PendingNotificationMap.cpp
Comment (gjc13) [reply src/app/clusters/bindings/PendingNotificationMap.cpp:27]: The binding table size will be typicall 10 to 20.

File: src/app/clusters/bindings/PendingNotificationMap.cpp
Comment (gjc13) [reply src/app/clusters/bindings/PendingNotificationMap.cpp:64]: The `pendingNotificationMap` works as a FIFO queue. When we fail to allocate a CASEClient we need to release an active outgoing CASE connection so we need to release the LRU entry from the pending notifications. That means the first peer in the queue.

File: src/app/clusters/bindings/PendingNotificationMap.cpp
Comment (gjc13) [reply src/app/clusters/bindings/PendingNotificationMap.cpp:92]: Realized that the `PendingNotificationMap` size will never be larger than the binding table. so we don't need `mNextToOverride` at all. Will merge the context and pending flag to the binding table storage.

File: src/app/clusters/bindings/PendingNotificationMap.cpp
Comment (bzbarsky-apple) [reply src/app/clusters/bindings/PendingNotificationMap.cpp:64]: OK.  Documenting _that_ would be very helpful in terms of understanding what this code is trying to accomplish.

File: src/app/clusters/bindings/PendingNotificationMap.cpp
Comment (gjc13) [reply src/app/clusters/bindings/PendingNotificationMap.cpp:99]: Yes, typically 50 bytes under the defaultc configuration.

File: src/app/clusters/bindings/PendingNotificationMap.h
Comment (gjc13) [reply src/app/clusters/bindings/PendingNotificationMap.h:48]: Since we won't have more entries than those in the binding table, `mNextToOverride` will be always zero. Will remove the member and the related logic.

File: src/app/clusters/bindings/BindingManager.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/BindingManager.cpp:108]: Might be simpler to just remove all entries that match the fabric/node of `device->GetPeerId()` so we don't have to save off `fabricToRemove`/`nodeToRemove`.

File: src/app/clusters/bindings/PendingNotificationMap.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/PendingNotificationMap.cpp:68]:
```suggestion
    uint16_t minLastAppearValue = UINT16_MAX;
```

File: src/app/clusters/bindings/PendingNotificationMap.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/PendingNotificationMap.cpp:27]:
```suggestion
    // When entries are added to PendingNotificationMap, they are appended to the end.
    // To find the LRU peer, we need to find the peer whose last entry in the map is closer
    // to the start of the list than the last entry of any other peer.

    // First, set up a way to easily track which entries correspond to the same peer.
    uint8_t bindingWithSamePeer[EMBER_BINDING_TABLE_SIZE];
```

**Review by bzbarsky-apple — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
Will need substantial changes and re-review once #14487 merges.

File: src/app/clusters/bindings/BindingManager.cpp
Comment (gjc13) [reply src/app/clusters/bindings/BindingManager.cpp:108]: The `PendingNotificationMap` stores only the fabric index. Mapping the `compressedFabricId` back to the fabric index requires iterating through the fabric table and is not supported now. Storing the index may be the simpler way.

**Comment by gjc13** [comment]
@bzbarsky-apple Merged latest master and comments resolved.

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**
